### PR TITLE
Settable shaders and other niceties

### PR DESCRIPTION
**Pull request by IceDragon200**
Renderable objects can now have their shaders set

Font, Sprite and Spritesheet now have a `#shader` and `#shader=` method,
the default shader for each class can be set via, `.default_shader=`,
all objects created after setting the default_shader will be affected,
objects created before will be unaffected.

TODO: Deprecate ShaderLoader

### CODE REVIEW

**Comment by IceDragon200** [comment]
@archSeer: Should we expose the `SetAttribute` / `SetUniform`?

**Comment by archSeer** [comment]
So here's my idea, `Shader.new(vert_str, frag_str)`. We completely remove file IO from the shader that way. Then in ruby, you can declare 
```ruby
def Shader.load(vert, frag)
  new(File.open(vert), File.open(frag))
end
```

File: modules/graphics/mrblib/font.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+module Moon
+  class Font
+    class << self
+      attr_accessor :default_shader
+    end
+
+    alias :initialize_wo_shader :initialize
```
Comment (archSeer) [inline modules/graphics/mrblib/font.rb]: All of this should be a mixin. Make a note to use `Module.prepend` in the future?

File: modules/graphics/mrblib/font.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    def initialize(*args, &block)
+      initialize_wo_shader(*args, &block)
+      post_initialize
+    end
+
+    def post_initialize
+      self.shader = self.class.default_shader
```
Comment (archSeer) [inline modules/graphics/mrblib/font.rb]: Here's a better idea:

```ruby
attr_accessor :shader

def shader
  @shader || @@default_shader
end
```

**Comment by archSeer** [comment]
And yes, those two should be exposed. You can add them now and I'll soon switch over, ideally I'm going to be porting Sprite and Spritesheet mainly into ruby.

**Comment by archSeer** [comment]
(Also holy shit, didn't know you got travis running! We could now run your test suite as well?)

File: modules/graphics/src/sprite.cxx [hunk 3]
```diff
@@ -66,27 +85,11 @@ namespace Moon {
...
-  };
-
-  void Sprite::SetClipRect(IntRect clip) {
-    m_clip_rect = std::move(clip); // passing by value already makes a copy
-    GenerateBuffers();
-  };
-
   void Sprite::Render(const float x, const float y, const float z) {
-    m_shader->Use();
+    if (!m_texture) return;
```
Comment (archSeer) [inline modules/graphics/src/sprite.cxx]: Why was this check removed?

File: modules/graphics/src/mrb_sprite.cxx [hunk 4]
```diff
@@ -136,13 +125,13 @@ sprite_render(mrb_state *mrb, mrb_value self)
 }
 
 static mrb_value
-sprite_get_opacity(mrb_state *mrb, mrb_value self)
+sprite_opacity_get(mrb_state *mrb, mrb_value self)
```
Comment (archSeer) [inline modules/graphics/src/mrb_sprite.cxx]: Lots of renames that should be a separate commit.

File: modules/graphics/mrblib/font.rb (hunk 1 above)
Comment (IceDragon200) [inline modules/graphics/mrblib/font.rb]: mruby doesn't have `Module#prepend`

File: modules/graphics/mrblib/font.rb (hunk 2 above)
Comment (IceDragon200) [inline modules/graphics/mrblib/font.rb]: The C side need to set the underlying structure's shader, which is why its initialized afterwards, though, I could try pulling it from the shader method using `mrb_funcall()`, nice idea!

File: modules/graphics/src/mrb_sprite.cxx (hunk 4 above)
Comment (IceDragon200) [inline modules/graphics/src/mrb_sprite.cxx]: True, I forgot to :x

File: modules/graphics/mrblib/font.rb (hunk 1 above)
Comment (archSeer) [inline modules/graphics/mrblib/font.rb]: Hence the "in the future" part. If you do restructuring as I suggest later, we don't need aliases at all.

File: modules/graphics/mrblib/display_object.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+module Moon
+  class DisplayObject
```
Comment (archSeer) [inline modules/graphics/mrblib/display_object.rb]: Nope. Something among the lines of:

```ruby
module Moon::Shadable

  def self.included(base)
    base.attr_accessor :shader
    base.class.attr_accessor :default_shader
  end

  def shader
    @shader || @@default_shader
  end
end

class Moon::Sprite
  include Shadable
end
```

File: modules/graphics/mrblib/font.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+module Moon
+  class Font < DisplayObject
```
Comment (archSeer) [inline modules/graphics/mrblib/font.rb]: There's a stray patch here, only on Font.

File: modules/graphics/src/shader.cxx [hunk 7]
```diff
@@ -7,13 +7,19 @@ namespace Moon {
...
   };
 
   Shader::~Shader() {
-    glDeleteProgram(m_program);
+    if (m_program) {
+      glDeleteProgram(m_program);
+    }
+  };
+
+  void Shader::CreateFromString(const std::string vertexContent, const std::string fragmentContent) {
```
Comment (archSeer) [inline modules/graphics/src/shader.cxx]: CreateFromString, CreateFromFile can now be removed and CreateProgram can be moved to init.

File: modules/graphics/src/sprite.cxx [hunk 8]
```diff
@@ -15,13 +15,32 @@ namespace Moon {
     origin = Vector2(0, 0);
     color = Vector4(1.0, 1.0, 1.0, 1.0);
     tone = Vector4(0.0, 0.0, 0.0, 1.0);
-    m_shader = ShaderLoader::GetQuadShader();
+    m_texture = NULL;
```
Comment (archSeer) [inline modules/graphics/src/sprite.cxx]: Aren't pointers initialized to null?

File: modules/graphics/src/renderable.cxx [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+  Renderable::Renderable() {
+    m_shader = NULL;
+  }
+
+  void Renderable::SetShader(Moon::Shader *shader) {
+    m_shader = shader;
+  }
+
+  Moon::Shader* Renderable::GetShader() {
+    return m_shader;
```
Comment (archSeer) [inline modules/graphics/src/renderable.cxx]: How is this any better than just directly accessing `m_shader`? Less indirections the better.

File: modules/graphics/src/sprite.cxx (hunk 8 above)
Comment (IceDragon200) [inline modules/graphics/src/sprite.cxx]: Dunno, and don't trust compilers.

File: modules/graphics/src/renderable.cxx (hunk 9 above)
Comment (archSeer) [inline modules/graphics/src/renderable.cxx]: The way it worked before was fine, it just needed to be initialized to `NULL`.

File: modules/graphics/src/shader.cxx [hunk 10]
```diff
@@ -7,13 +7,19 @@ namespace Moon {
...
-  Shader::Shader(const char* vertexfile, const char* fragmentfile)
+  Shader::Shader()
   {
-    m_program = CreateProgram(vertexfile, fragmentfile);
+    m_program = 0;
   };
 
   Shader::~Shader() {
-    glDeleteProgram(m_program);
+    if (m_program) {
```
Comment (archSeer) [inline modules/graphics/src/shader.cxx]: Unnecessary check. `m_program` will always be true (because we use an int, not a pointer), and in the case of 0,`glDeleteProgram(0)` will just be silent.

File: modules/graphics/src/renderable.cxx (hunk 9 above)
Comment (IceDragon200) [inline modules/graphics/src/renderable.cxx]: m_shader is private, which is why you can't set it, from the child class, blame C++.

**Comment by archSeer** [comment]
Good job! :+1:
